.builders/upload.py

```diff
@@ -69,96 +76,140 @@ def hash_file(path: Path) -> str:
-                        print(f'{prefix} {project_name}=={project_metadata["Version"]} already exists '
-                              'with the same hash')
-                        continue
-
-                build_number = timestamp_build_number()
-                artifact_name = f'{name}-{version}-{build_number}-{python_tag}-{abi_tag}-{platform_tag}.whl'
-                artifact = bucket.blob(f'{artifact_type}/{project_name}/{artifact_name}')
-
-            print(f'{padding}Artifact: {artifact_name}')
-            artifact.upload_from_filename(str(wheel))
-
-            requires_python = project_metadata.get('Requires-Python', '').replace('<', '&lt;').replace('>', '&gt;')
-
-            artifact.metadata = {'requires-python': requires_python, 'sha256': sha256_digest}
-            artifact.patch()
+                artifact_type = entry.name
+                artifact_types.add(artifact_type)
+                display_message_block(f'Processing {artifact_type} wheels')
+
+                upload_data: list[tuple[str, email.Message, Path]] = []
+                for wheel in entry.iterdir():
+                    project_metadata = extract_metadata(wheel)
+                    project_name = project_metadata['Name']
+                    if not is_valid_project_name(project_name):
+                        message = f'Invalid project name `{project_name}` found in wheel: {wheel.name}'
+                        raise RuntimeError(message)
+
+                    upload_data.append((normalize_project_name(project_name), project_metadata, wheel))
+
+                queued = len(upload_data)
+                upload_data.sort()
+
+                for i, (project_name, project_metadata, wheel) in enumerate(upload_data, start=1):
+                    prefix = f'({i}/{queued})'
+                    padding = ' ' * (len(prefix) + 1)
+                    print(f'{prefix} Name: {project_metadata["Name"]}')
+                    print(f'{padding}Version: {project_metadata["Version"]}')
+
+                    sha256_digest = hash_file(wheel)
+                    index_url = f'{STORAGE_URL}/{artifact_type}/{project_name}'
+                    already_uploaded = False
+                    dep_entry = ""
+
+                    if artifact_type == 'external':
+                        artifact_name = wheel.name
+                        # https://agent-int-packages.datadoghq.com/external/cffi/cffi-1.17.1-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
+
+                        artifact = bucket.blob(f'{artifact_type}/{project_name}/{artifact_name}')
+
+                        # PyPI artifacts never change, so we don't need to upload them again.
+                        if artifact.exists():
+                            print(f'{prefix} {project_name}=={project_metadata["Version"]} already exists')
+                            already_uploaded = True
+                            dep_entry = f'{project_name} @ {artifact.name}'
```

Comment:
Do we need this here? I feel we do not need to add the `already_uploaded` variable. The script is already complex enough to keep one more extra variable through the iteration. The working memory at this point is probably too much.

Instead of keeping the `dep_entry` for all cases and not continue to add the line at the end, we can just call append before continuing.

We can do this here and in any other place below: 


```suggestion
                            lockfile_lines.append(f'{project_name} @ {artifact.name}')
                            continue
```

Reply:
Good idea, I updated the logic. Would love if you could double check if the new logic makes sense 🙏